### ISSUE TITLE
How to model a 'Phase' or a 'Step' in a public service? (SDGR, art. 6 & 10)

### ISSUE THREAD

**Issue by kareljanr** [catalogue-of-services-isa/SDG-model#3]
**General**
Based on [Issue 31](https://github.com/catalogue-of-services-isa/CPSV-AP/issues/31) and as the SDG Regulation discusses 'procedural steps' in article 6 and 10, we propose to model this concept in the SDG domain model.

**Should we add this concept?**
- Required (article 6 & 10, SDGR)
- Useful for citizens (see what steps there are in a procedure)
- Useful for citizens (feedback on a specific kind of step)
- Hard to align with the local level - e.g. 8000+ municipalities in Spain with other procedural steps

**Approaches**
1. This concept could be added to the model as a property of the Public Service class. A controlled vocabulary / classification could be added, e.g. the Spanish example:
A. Initiation
A.1. Access
A.2. Fulfilment
A.3. Submission
B. Processing
B.1. Acceptance
B.2. Acts
C. Finalization
C.1. Resolution
C.2. Delivery

2. As commented in [Issue 42](https://github.com/catalogue-of-services-isa/CPSV-AP/issues/42), a 'Procedure' could be linked to 0:* Procedure(s). A possible approach could be to add a relationship 'follows' to Procedure that points to itself. In this example one procedure follows another procedure, which means that a procedural step would equal a procedure.

3. Other approaches could be possible. We depend on your input to identify other possible approaches or to improve the identified approaches. Feel free to provide your opinion on this issue.

**Comment by Marco-ICTU-NL** [comment]
The way it is modeled now (loose actions that can 'follow' each other) is IMHO too limited. Procedure steps (the term used in the SDGR) are not always sequential. There may be optional steps that can be skipped, there may be conditional steps based on the situation, etc. I think the best solution here would be to model the procedural steps as a free text description. After all, the aim is to provide citizens with information they can read, not necessarily with a (complicated) model that suits all situations

**Comment by giorgialodi** [comment]
I agree with @Marco-ICTU-NL on the fact that probably the model of actions (BTW: are they public actions?) is limited.  However, if we need to consider all the possible cases we should foresee them. And using free text is not very nice: we should pass from strings to codes :) at least in RDF data models! Because even on that data we may enable interesting analysis.
It is true that the property "follows" indicates that an action comes after another but I am wondering:
1. are there cases in which a step is required for executing the following one? Said in another way, are there cases in which actions must be executed as an ordered sequence? If so, "follows" only is not sufficient and we should model an ordered sequence pattern
2. are the conditions based on the situation a kind of criterion? 
3. if some steps are optional, I think this should be modeled in some way so that to mark the step as optional in the description of the overall public service. 
4. are the input and output concepts more related to the actions rather than the public service? For executing an action I need to provide an input and the action produces the output (that may be the input of the next action) and so on.

**Comment by Marco-ICTU-NL** [comment]
We need to keep in mind that the data in question has to be delivered by all kinds of authorities,  including a great number of municipalities. To map their data (often free text for a webpage) to a very detailed structure means a lot of work and a lot of money. We have 355 municipalities, and other countries have many more.

**Comment by giorgialodi** [comment]
@Marco-ICTU-NL I know the problem. In my country (Italy) there are almost 8.000 municipalities :) and I understand your point. If we construct the model in such a way to have both possibilities we are even more flexible: the ones who want to take advantage of a more detailed model can exploit it, the others who wants to apply a more lightweight model can use it anyway.

**Comment by barthelemyf** [comment]
## Proposed resolution
The class Action was foreseen as a generic approach which could fit the different cases pointed out above:

- **Optional vs required**: we could add an optional property to capture this.
- **Unconditional vs conditional**: we could add an optional property to capture this.
- **Executed by** a **Person** in the role of a citizen, by an **Organisation** in the role of the Service provider or any other Agent and/or Role combination

With the relation “follows” and its cardinality, we leave it up to the public administrations to decide if they would like to detail extensively in separated actions the execution of the service or if a single action is sufficient to describe from a high level the execution of the service. This is the flexibility provided by the property _description_ with a free text field (in this case, this would include all activities in one description for the so-called above “procedural steps”).
Regarding the sequence, the following visual shows how the relation “follows” could work for different cases together with the optional and conditional properties and the relation executes pointing to an agent:
<img width="719" alt="Sequence of actions" src="https://user-images.githubusercontent.com/32893766/74884332-385a2e00-5373-11ea-86c0-752338d39a75.png">
Each action follows zero to multiple action(s). Together with conditions expressed as part of the criterion, the sequence of actions could be recreated easily by a machine thanks to the "follows" relation. Please let us know if you identified cases which would not work.
In our understanding, the conditions and their related input and output could indeed be directly linked to the class Action and not to the class Public Service as it is currently modelled. The conditions could be modelled according to the class Criterion and related class from the CCCEV. The input could be represented by the class Evidence. Finally the class Action could produce an Output which could be reused by the next action (the latest action could then produce the final output to be delivered to the requestor).
